## Project 1b: Language Modeling

In this project, you will implement several different types of language models for text. We'll start with n-gram models, then move on to neural n-gram and LSTM language models.

Warning: Do not start this project the day before it is due! Some parts require 20 minutes or more to run, so debugging and tuning can take a significant amount of time.

Our dataset for this project will be the WikiText2 language modeling dataset. This dataset comes with some of the basic preprocessing done for us, such as tokenization and rare word filtering (using the <unk> token). Therefore, we can assume that all word types in the test set also appear at least once in the training set. We'll also use the torchtext library to help with some of the data preprocessing, such as converting tokens into id numbers.

Some of the functions below require an older version of torchtext (`0.10.0`) than the default one Kaggle gives you.

In [1]:
# This block handles some basic setup and data loading.  
# You shouldn't need to edit this, but if you want to 
# import other standard python packages, that is fine.

# imports
from collections import defaultdict, Counter
import numpy as np
import os
import math
import tqdm
import random
import pdb

import torch
from torch import nn
import torch.nn.functional as F
import torchtext

from torchtext.legacy import data
from torchtext.legacy import datasets

# download and load the data
text_field = data.Field()
datasets = datasets.WikiText2.splits(root='.', text_field=text_field)

train_dataset, validation_dataset, test_dataset = datasets

text_field.build_vocab(train_dataset, validation_dataset, test_dataset)
vocab = text_field.vocab
vocab_size = len(vocab)

train_text = train_dataset.examples[0].text # a list of tokens (strings)
validation_text = validation_dataset.examples[0].text

print(validation_text[:30])

['<eos>', '=', 'Homarus', 'gammarus', '=', '<eos>', '<eos>', 'Homarus', 'gammarus', ',', 'known', 'as', 'the', 'European', 'lobster', 'or', 'common', 'lobster', ',', 'is', 'a', 'species', 'of', '<unk>', 'lobster', 'from', 'the', 'eastern', 'Atlantic', 'Ocean']


In [2]:
text_field.vocab.freqs['.']

90077

In [ ]:
os.makedirs('./saves/', exist_ok=True)

In [3]:
def check_validity(model):
    """Performs several sanity checks on your model:
    1) That next_word_probabilities returns a valid distribution
    2) That perplexity matches a perplexity calculated from next_word_probabilities

    Although it is possible to calculate perplexity from next_word_probabilities, 
    it is still good to have a separate more efficient method that only computes 
    the probabilities of observed words.
    """

    log_probabilities = []
    for i in range(10):
        prefix = validation_text[:i]
        probs = model.next_word_probabilities(prefix)
        assert min(probs) >= 0, "Negative value in next_word_probabilities"
        assert max(probs) <= 1 + 1e-8, "Value larger than 1 in next_word_probabilities"
        assert abs(sum(probs)-1) < 1e-4, "next_word_probabilities do not sum to 1"

        word_id = vocab.stoi[validation_text[i]]
        selected_prob = probs[word_id]
        log_probabilities.append(math.log(selected_prob))

    perplexity = math.exp(-np.mean(log_probabilities))
    your_perplexity = model.perplexity(validation_text[:10])
    assert abs(perplexity-your_perplexity) < 0.1, "your perplexity does not " + \
    "match the one we calculated from `next_word_probabilities`,\n" + \
    "at least one of `perplexity` or `next_word_probabilities` is incorrect.\n" + \
    f"we calcuated {perplexity} from `next_word_probabilities`,\n" + \
    f"but your perplexity function returned {your_perplexity} (on a small sample)."

We've implemented a unigram model here as a demonstration.

In [4]:
class UnigramModel:
    def __init__(self, train_text):
        self.counts = Counter(train_text)
        self.total_count = len(train_text)

    def probability(self, word):
        return self.counts[word] / self.total_count

    def next_word_probabilities(self, text_prefix):
        """Return a list of probabilities for each word in the vocabulary."""
        return [self.probability(word) for word in vocab.itos]

    def perplexity(self, full_text):
        """Return the perplexity of the model on a text as a float.
        
        full_text -- a list of string tokens
        """
        log_probabilities = []
        for word in full_text:
            # Note that the base of the log doesn't matter 
            # as long as the log and exp use the same base.
            log_probabilities.append(math.log(self.probability(word), 2))
        return 2 ** -np.mean(log_probabilities)

unigram_demonstration_model = UnigramModel(train_text)
print('unigram validation perplexity:', 
      unigram_demonstration_model.perplexity(validation_text))
check_validity(unigram_demonstration_model)

unigram validation perplexity: 965.0860734119312


To generate from a language model, we can sample one word at a time conditioning on the words we have generated so far.

In [5]:
def generate_text(model, n=20, prefix=('<eos>', '<eos>')):
    prefix = list(prefix)
    for _ in range(n):
        probs = model.next_word_probabilities(prefix)
        word = random.choices(vocab.itos, probs)[0]
        prefix.append(word)
    return ' '.join(prefix)

print(generate_text(unigram_demonstration_model))

<eos> <eos> she head in " been division the town League = The a " The were Uyghur great on <unk> From


In fact there are many strategies to get better-sounding samples, such as only sampling from the top-k words or sharpening the distribution with a temperature.  You can read more about sampling from a language model in this recent paper: https://arxiv.org/pdf/1904.09751.pdf.

In [6]:
%%capture
!wget -N https://cal-cs288.github.io/sp21/project_files/proj_1/eval_prefixes.txt -P ./data/
!wget -N https://cal-cs288.github.io/sp21/project_files/proj_1/eval_output_vocab.txt -P ./data/
!wget -N https://cal-cs288.github.io/sp21/project_files/proj_1/eval_prefixes_short.txt -P ./data/
!wget -N https://cal-cs288.github.io/sp21/project_files/proj_1/eval_output_vocab_short.txt -P ./data/

You will need to submit some outputs from the models you implement for us to grade.  The following function will be used to generate the required output files.

In [7]:
def save_truncated_distribution(model, filename, short=True):
    """Generate a file of truncated distributions.
    
    Probability distributions over the full vocabulary are large,
    so we will truncate the distribution to a smaller vocabulary.

    Please do not edit this function
    """
    vocab_name = './data/eval_output_vocab'
    prefixes_name = './data/eval_prefixes'

    if short: 
      vocab_name += '_short'
      prefixes_name += '_short'

    with open('{}.txt'.format(vocab_name), 'r') as eval_vocab_file:
        eval_vocab = [w.strip() for w in eval_vocab_file]
    eval_vocab_ids = [vocab.stoi[s] for s in eval_vocab]

    all_selected_probabilities = []
    with open('{}.txt'.format(prefixes_name), 'r') as eval_prefixes_file:
        lines = eval_prefixes_file.readlines()
        for line in tqdm.notebook.tqdm(lines, leave=False):
            prefix = line.strip().split(' ')
            probs = model.next_word_probabilities(prefix)
            selected_probs = np.array([probs[i] for i in eval_vocab_ids], dtype=np.float32)
            all_selected_probabilities.append(selected_probs)

    all_selected_probabilities = np.stack(all_selected_probabilities)
    np.save(filename, all_selected_probabilities)
    print('saved', filename)

In [8]:
save_truncated_distribution(unigram_demonstration_model, 
                            './saves/unigram_demonstration_predictions.npy')

  0%|          | 0/1000 [00:00<?, ?it/s]

saved ./saves/unigram_demonstration_predictions.npy


### N-gram Model

Now it's time to implement an n-gram language model.

Because not every n-gram will have been observed in training, use add-alpha smoothing to make sure no output word has probability 0.

$$P(w_2|w_1)=\frac{C(w_1,w_2)+\alpha}{C(w_1)+N\alpha}$$

where $N$ is the vocab size and $C$ is the count for the given bigram.  An alpha value around `3e-3`  should work.  Later, we'll replace this smoothing with model backoff.

One edge case you will need to handle is at the beginning of the text where you don't have `n-1` prior words.  You can handle this however you like as long as you produce a valid probability distribution, but just using a uniform distribution over the vocabulary is reasonable for the purposes of this project.

A properly implemented bi-gram model should get a perplexity below 510 on the validation set.

**Note**: Do not change the signature of the `next_word_probabilities` and `perplexity` functions.  We will use these as a common interface for all of the different model types.  Make sure these two functions call `n_gram_probability`, because later we are going to override `n_gram_probability` in a subclass. 
Also, we suggest pre-computing and caching the counts $C$ when you initialize `NGramModel` for efficiency. 

In [9]:
import nltk

def compute_freq(tokens, n_value=2):
    assert n_value >= 0
    if n_value == 0:
        return {(): len(tokens)}

    ngrams = nltk.ngrams(tokens, n_value)
    ngram_fdist = nltk.FreqDist(ngrams)
    return ngram_fdist

In [10]:
class NGramModel(UnigramModel):
    def __init__(self, train_text, n=2, alpha=3e-3):
        super().__init__(train_text)

        # get counts and perform any other setup
        self.n = n
        self.smoothing = alpha

        # YOUR CODE HERE

        self.ngram_counts = compute_freq(train_text, n)
        self.base_counts = compute_freq(train_text, n-1)

    def n_gram_probability(self, n_gram):
        """Return the probability of the last word in an n-gram.
        
        n_gram -- a list of string tokens
        returns the conditional probability of the last token given the rest.
        """
        assert len(n_gram) == self.n

        # YOUR CODE HERE

        return (self.ngram_counts[tuple(n_gram)] + self.smoothing) / (
                self.base_counts[tuple(n_gram[:-1])] + len(vocab) * self.smoothing)

    def next_word_probabilities(self, text_prefix):
        """Return a list of probabilities for each word in the vocabulary."""

        # YOUR CODE HERE
        # use your function n_gram_probability
        # vocab.itos contains a list of words to return probabilities for

        if len(text_prefix) < (self.n-1):
            return [self.probability(word) for word in vocab.itos]

        prefix = text_prefix[(1-self.n):] if self.n > 1 else []
        return [self.n_gram_probability(prefix+[word]) for word in vocab.itos]

    def perplexity(self, full_text):
        """ full_text is a list of string tokens
        return perplexity as a float """

        # YOUR CODE HERE
        # use your function n_gram_probability
        # This method should differ a bit from the example unigram model because 
        # the first n-1 words of full_text must be handled as a special case.

        log_probabilities = []
        for i in range(self.n-1):
            word = full_text[i]
            log_probabilities.append(math.log(self.probability(word), 2))
        for i in range(self.n-1, len(full_text)):
            n_gram = full_text[(i+1-self.n):(i+1)]
            log_probabilities.append(math.log(self.n_gram_probability(n_gram), 2))
        return 2 ** -np.mean(log_probabilities)

unigram_model = NGramModel(train_text, 1)
check_validity(unigram_model)
print('unigram validation perplexity:', unigram_model.perplexity(validation_text)) # this should be the almost the same as our unigram model perplexity above

bigram_model = NGramModel(train_text, n=2)
check_validity(bigram_model)
print('bigram validation perplexity:', bigram_model.perplexity(validation_text))

trigram_model = NGramModel(train_text, n=3)
check_validity(trigram_model)
print('trigram validation perplexity:', trigram_model.perplexity(validation_text)) # this won't do very well...

save_truncated_distribution(bigram_model, './saves/bigram_predictions.npy') # this might take a few minutes

unigram validation perplexity: 965.0913686618096
bigram validation perplexity: 504.4148539982192
trigram validation perplexity: 2965.494861349281


  0%|          | 0/1000 [00:00<?, ?it/s]

saved ./saves/bigram_predictions.npy


Please download `bigram_predictions.npy` once you finish this section so that you can submit it.

In the block below, please report your bigram validation perplexity.  (We will use this to help us calibrate our scoring on the test set.)

<!-- Do not remove this comment, it is used by the autograder: RqYJKsoTS6 -->

Bigram validation perplexity: ***504.4148539982192***

We can also generate samples from the model to get an idea of how it is doing.

In [11]:
print(generate_text(bigram_model))

<eos> <eos> = = = Interior aunt reverting founder Precautions Whitney Obsessions Katia eastwards VIII surveys Mearns Creation and N. camera for


We now free up some RAM, **it is important to run the cell below, otherwise you will likely run out of RAM in the Colab runtime.**

In [12]:
# Free up some RAM. 
del bigram_model
del trigram_model

This basic model works okay for bigrams, but a better strategy (especially for higher-order models) is to use backoff.  Implement backoff with absolute discounting.
$$P\left(w_i|w_{i-n+1}^{i-1}\right)=\frac{max\left\{C(w_{i-n+1}^i)-\delta,0\right\}}{\sum_{w_i} C(w_{i-n+1}^i)} + \alpha(w_{i-n+1}^{i-1}) P(w_i|w_{i-n+2}^{i-1})$$

$$\alpha\left(w_{i-n+1}^{i-1}\right)=\frac{\delta N_{1+}(w_{i-n+1}^{i-1})}{{\sum_{w_i} C(w_{i-n+1}^i)}}$$
where $N_{1+}$ is the number of words that appear after the previous $n-1$ words (the number of times the max will select something other than 0 in the first equation).  If $\sum_{w_i} C(w_{i-n+1}^i)=0$, use the lower order model probability directly (the above equations would have a division by 0).

We found a discount $\delta$ of 0.9 to work well based on validation performance.  A trigram model with this discount value should get a validation perplexity below 275.

In [13]:
class DiscountBackoffModel(NGramModel):
    def __init__(self, train_text, lower_order_model, n=2, alpha=3e-3, delta=0.9):
        # get counts and perform any other setup
        self.n = n
        self.smoothing = alpha
        self.lower_order_model = lower_order_model
        self.discount = delta

        # YOUR CODE HERE

        self.counts = Counter(train_text)
        self.total_count = len(train_text)

        self.ngram_counts = nltk.FreqDist()
        for i in range(len(train_text)-n):
            prefix = tuple(train_text[i:(i+n-1)])
            if prefix not in self.ngram_counts:
                self.ngram_counts[prefix] = nltk.FreqDist()
            last = train_text[i+n-1]
            if last not in self.ngram_counts[prefix]:
                self.ngram_counts[prefix][last] = 1
            else:
                self.ngram_counts[prefix][last] += 1
        self.base_counts = compute_freq(train_text, n-1)

    def n_gram_probability(self, n_gram):
        assert len(n_gram) == self.n

        # YOUR CODE HERE
        # back off to the lower_order model with n'=n-1 using its n_gram_probability function

        lower_order_prob = self.lower_order_model.n_gram_probability(n_gram[1:])

        ngram = tuple(n_gram)
        last = ngram[-1]
        prefix = ngram[:-1]

        prefix_counts = self.base_counts[prefix]
        if prefix_counts == 0:
            return lower_order_prob

        alpha = self.discount * len(self.ngram_counts[prefix]) / prefix_counts

        return (max([self.ngram_counts[prefix][ngram[-1]] - self.discount, 0]) / prefix_counts + 
            alpha * lower_order_prob
        )

bigram_backoff_model = DiscountBackoffModel(train_text, unigram_model, 2)
trigram_backoff_model = DiscountBackoffModel(train_text, bigram_backoff_model, 3)
check_validity(trigram_backoff_model)
print('trigram backoff validation perplexity:', trigram_backoff_model.perplexity(validation_text))

trigram backoff validation perplexity: 271.10633787343585


Fill in your trigram backoff perplexity here.

<!-- Do not remove this comment, it is used by the autograder: RqYJKsoTS6 -->

Trigram backoff validation perplexity: ***271.10633787343585***



Free up RAM. 

In [14]:
# Release models we don't need any more. 
del unigram_model
del bigram_backoff_model
del trigram_backoff_model

### Neural N-gram Model

In this section, you will implement a neural version of an n-gram model.  The model will use a simple feedforward neural network that takes the previous `n-1` words and outputs a distribution over the next word.

You will use PyTorch to implement the model.  We've provided a little bit of code to help with the data loading using PyTorch's data loaders (https://pytorch.org/docs/stable/data.html)

In [15]:
if torch.cuda.is_available():
    print('Found GPU')
    device = 'cuda'
else:
    print('Did not find GPU')
    device = 'cpu'


class Module(nn.Module):
    def __init__(self):
        super().__init__()

    def move_input(self, input):
        """
        Move minibatch tensors to CPU/GPU.
        """
        if isinstance(input, list):
            return [i.to(self.device) if i is not None else None for i in input]
        elif isinstance(input, tuple):
            return tuple([i.to(self.device) if i is not None else None for i in input])
        else:
            return input.to(self.device)

    def move_inputs(self, *inputs: torch.Tensor):
        """
        Move minibatch tensors to CPU/GPU.
        """
        return [self.move_input(i) if i is not None else None for i in inputs]

    def to_device(self, device):
        self.device = device
        self.to(self.device)

Found GPU


In [16]:
def ids(tokens):
    return [vocab.stoi[t] for t in tokens]


class NeuralNgramDataset(torch.utils.data.Dataset):
    def __init__(self, text_token_ids, n):
        self.text_token_ids = text_token_ids
        self.n = n

    def __len__(self):
        return len(self.text_token_ids)

    def __getitem__(self, i):
        if i < self.n-1:
            prev_token_ids = [vocab.stoi['<eos>']] * (self.n-i-1) + self.text_token_ids[:i]
        else:
            prev_token_ids = self.text_token_ids[i-self.n+1:i]

        assert len(prev_token_ids) == self.n-1

        x = torch.tensor(prev_token_ids)
        y = torch.tensor(self.text_token_ids[i])
        return x, y

A model with the following architecture and hyperparameters should reach a validation perplexity below 226.
* embed the words with dimension 128, then flatten into a single embedding for $n-1$ words (with size $(n-1)*128$)
* run 2 hidden layers with 1024 hidden units, then project down to size 128 before the final layer (ie. 4 layers total). 
* use weight tying for the embedding and final linear layer (this made a very large difference in our experiments); you can do this by creating the output layer with `nn.Linear`, then using `F.embedding` with the linear layer's `.weight` to embed the input
* rectified linear activation (ReLU) and dropout 0.1 after first 2 hidden layers. **Note: You will likely find a performance drop if you add a nonlinear activation function after the dimension reduction layer.**
* train for 10 epochs with the Adam optimizer (should take around 15-20 minutes)
* do early stopping based on validation set perplexity (see Project 1a)


We encourage you to try other architectures and hyperparameters, and you will likely find some that work better than the ones listed above.  A proper implementation with these should be enough to receive full credit on the assignment, though.

In [19]:
class NeuralNGramNetwork(Module):
    # a PyTorch Module that holds the neural network for your model

    def __init__(self, n, device):
        super().__init__()
        self.n = n

        # YOUR CODE HERE

        self.hidden_network = nn.Sequential(
            nn.Linear(128*(n-1), 1024),
            nn.LeakyReLU(),
            nn.Dropout(0.1),
            nn.Linear(1024, 1024),
            nn.LeakyReLU(),
            nn.Dropout(0.1),
            nn.Linear(1024, 128)
        )
        self.final_layer = nn.Linear(128, len(vocab))

        self.to_device(device)

    def forward(self, x):
        # x is a tensor of inputs with shape (batch, n-1)
        # this function returns a tensor of log probabilities with shape (batch, vocab_size)

        # YOUR CODE HERE

        x = self.move_input(x)

        x = F.embedding(x, self.final_layer.weight)
        x = x.view(x.shape[0], -1)
        return self.final_layer(self.hidden_network(x))


class NeuralNGramModel:
    # a class that wraps NeuralNGramNetwork to handle training and evaluation
    # it's ok if this doesn't work for unigram modeling
    def __init__(self, n, device):
        self.n = n
        self.network = NeuralNGramNetwork(n, device)

    def train(self, n_epochs=10):
        dataset = NeuralNgramDataset(ids(train_text), self.n)
        train_loader = torch.utils.data.DataLoader(dataset, batch_size=128, shuffle=True)
        # iterating over train_loader with a for loop will return a 2-tuple of batched tensors
        # the first tensor will be previous token ids with size (batch, n-1),
        # and the second will be the current token id with size (batch, )
        # you will need to move these tensors to GPU, e.g. by using the Tensor.cuda() function.

        # this will take some time to run; use tqdm.notebook.tqdm to get a progress bar 
        # (see Project 1a for example)

        # YOUR CODE HERE

        optimizer = torch.optim.Adam(self.network.parameters(), lr=1e-3)

        best_perplexity = float('inf')
        for epoch in range(n_epochs):
            print('Epoch', epoch)
            for prefix, last in tqdm.notebook.tqdm(train_loader, leave=False):
                logits = self.network(prefix)
                loss = F.cross_entropy(logits, last.to(logits.device))

                optimizer.zero_grad()
                loss.backward()
                optimizer.step()
        
            perplexity = self.perplexity(validation_text)
            print('Validation score:', perplexity)

            if perplexity < best_perplexity:
                best_perplexity = perplexity
                torch.save(self.network.state_dict(), './saves/best_NeuralNGramModel.pt')

        self.network.load_state_dict(torch.load('./saves/best_NeuralNGramModel.pt'))

    def next_word_probabilities(self, text_prefix):
        # YOUR CODE HERE
        # don't forget self.network.eval()
        # you will need to convert text_prefix from strings to numbers with the `ids` function
        # if your `perplexity` function below is based on a NeuralNgramDataset DataLoader, you will need to use the same strategy for prefixes with less than n-1 tokens to pass the validity check
        #   the data loader appends extra "<eos>" (end of sentence) tokens to the start of the input so there are always enough to run the network

        if len(text_prefix) < (self.n-1):
            text_prefix = ['<eos>'] * (self.n-len(text_prefix)-1) + text_prefix

        prefix = text_prefix[(1-self.n):]
        x = torch.tensor(ids(prefix)).view(1, -1)

        with torch.no_grad():
            self.network.eval()

            logits = self.network(x)
            probs = F.softmax(logits, dim=-1)

            self.network.train()

        return probs.squeeze(0).cpu().numpy()

    def perplexity(self, text):
        # you may want to use a DataLoader here with a NeuralNgramDataset
        # don't forget self.network.eval()

        # YOUR CODE HERE

        dataset = NeuralNgramDataset(ids(text), self.n)
        eval_loader = torch.utils.data.DataLoader(dataset, batch_size=128)

        log_probabilities = []
        with torch.no_grad():
            self.network.eval()

            for prefix, last in tqdm.notebook.tqdm(eval_loader, leave=False):
                logits = self.network(prefix)
                probs = F.softmax(logits, dim=-1)

                selected_probs = probs[torch.arange(last.shape[0]), last]
                log_probabilities.append(torch.log2(selected_probs).cpu().numpy())

            self.network.train()

        return 2 ** -np.mean(np.concatenate(log_probabilities))

neural_trigram_model = NeuralNGramModel(3, device)
check_validity(neural_trigram_model)
neural_trigram_model.train()
print('neural trigram validation perplexity:', neural_trigram_model.perplexity(validation_text))

save_truncated_distribution(neural_trigram_model, './saves/neural_trigram_predictions.npy', short=False)

  0%|          | 0/1 [00:00<?, ?it/s]

Epoch 0


  0%|          | 0/16318 [00:00<?, ?it/s]

  0%|          | 0/1701 [00:00<?, ?it/s]

Validation score: 230.43698144590743
Epoch 1


  0%|          | 0/16318 [00:00<?, ?it/s]

  0%|          | 0/1701 [00:00<?, ?it/s]

Validation score: 213.18193954226751
Epoch 2


  0%|          | 0/16318 [00:00<?, ?it/s]

  0%|          | 0/1701 [00:00<?, ?it/s]

Validation score: 208.09478661635566
Epoch 3


  0%|          | 0/16318 [00:00<?, ?it/s]

  0%|          | 0/1701 [00:00<?, ?it/s]

Validation score: 206.6992810778369
Epoch 4


  0%|          | 0/16318 [00:00<?, ?it/s]

  0%|          | 0/1701 [00:00<?, ?it/s]

Validation score: 213.40605131432505
Epoch 5


  0%|          | 0/16318 [00:00<?, ?it/s]

  0%|          | 0/1701 [00:00<?, ?it/s]

Validation score: 215.97888079662343
Epoch 6


  0%|          | 0/16318 [00:00<?, ?it/s]

  0%|          | 0/1701 [00:00<?, ?it/s]

Validation score: 218.24942483144432
Epoch 7


  0%|          | 0/16318 [00:00<?, ?it/s]

  0%|          | 0/1701 [00:00<?, ?it/s]

Validation score: 221.28561122369618
Epoch 8


  0%|          | 0/16318 [00:00<?, ?it/s]

  0%|          | 0/1701 [00:00<?, ?it/s]

Validation score: 226.7833625399887
Epoch 9


  0%|          | 0/16318 [00:00<?, ?it/s]

  0%|          | 0/1701 [00:00<?, ?it/s]

Validation score: 227.82599213032614


  0%|          | 0/1701 [00:00<?, ?it/s]

neural trigram validation perplexity: 206.6992810778369


  0%|          | 0/5000 [00:00<?, ?it/s]

saved ./saves/neural_trigram_predictions.npy


Fill in your neural trigram perplexity.

<!-- Do not remove this comment, it is used by the autograder: RqYJKsoTS6 -->

Neural trigram validation perplexity: ***206.6992810778369***

Free up RAM.

In [45]:
# Delete model we don't need. 
del neural_trigram_model

### LSTM Model

For this stage of the project, you will implement an LSTM language model.

For recurrent language modeling, the data batching strategy is a bit different from what is used in some other tasks.  Sentences are concatenated together so that one sentence starts right after the other, and an unfinished sentence will be continued in the next batch.  We'll use the `torchtext` library to manage this batching for you.  To properly deal with this input format, you should save the last state of the LSTM from a batch to feed in as the first state of the next batch.  When you save state across different batches, you should call `.detach()` on the state tensors before the next batch to tell PyTorch not to backpropagate gradients through the state into the batch you have already finished (which will cause a runtime error).

We expect your model to reach a validation perplexity below 130.  The following architecture and hyperparameters should be sufficient to get there.
* 3 LSTM layers with 512 units
* dropout of 0.5 after each LSTM layer
* instead of projecting directly from the last LSTM output to the vocabulary size for softmax, project down to a smaller size first (e.g. 512->128->vocab_size). **NOTE: You may find that adding nonlinearities between these layers can hurt performance, try without first.**
* use the same weights for the embedding layer and the pre-softmax layer; dimension 128
* train with Adam (using default learning rates) for at least 20 epochs


In [18]:
class LSTMNetwork(Module):
    # a PyTorch Module that holds the neural network for your model

    def __init__(self, device):
        super().__init__()

        # YOUR CODE HERE

        self.lstm_network = nn.LSTM(128, 512, num_layers=3, dropout=0.5)
        self.hidden_network = nn.Sequential(
            nn.Dropout(0.5),
            nn.Linear(512, 128)
        )
        self.final_layer = nn.Linear(128, len(vocab))

        self.to_device(device)

    def forward(self, x, state):
        """Compute the output of the network.
        
        Note: In the Pytorch LSTM tutorial, the state variable is named "hidden":
        https://pytorch.org/tutorials/beginner/nlp/sequence_models_tutorial.html

        The torch.nn.LSTM documentation is quite helpful:
        https://pytorch.org/docs/stable/nn.html#lstm
    
        x - a tensor of int64 inputs with shape (seq_len, batch)
        state - a tuple of two tensors with shape (num_layers, batch, hidden_size)
                representing the hidden state and cell state of the of the LSTM.
        returns a tuple with two elements:
          - a tensor of log probabilities with shape (seq_len, batch, vocab_size)
          - a state tuple returned by applying the LSTM.
        """

        # Note that the nn.LSTM module expects inputs with the sequence 
        # dimension before the batch by default.
        # In this case the dimensions are already in the right order, 
        # but watch out for this since sometimes people put the batch first

        # YOUR CODE HERE

        x, state = self.move_inputs(x, state)

        x = F.embedding(x, self.final_layer.weight)
        x, state = self.lstm_network(x, state)
        return self.final_layer(self.hidden_network(x)), tuple([s.detach() for s in state])


class LSTMModel:
    "A class that wraps LSTMNetwork to handle training and evaluation."

    def __init__(self, device):
        self.network = LSTMNetwork(device)

    def train(self, n_epochs=20):
        train_iterator = data.BPTTIterator(
            train_dataset, batch_size=64, bptt_len=32, device=self.network.device)
        # Iterate over train_iterator with a for loop to get batches
        # each batch object has a .text and .target attribute with
        # token id tensors for the input and output respectively.

        # The initial state passed into the LSTM should be set to zero.

        # YOUR CODE HERE

        optimizer = torch.optim.Adam(self.network.parameters(), lr=1e-3)

        best_perplexity = float('inf')
        for epoch in range(n_epochs):
            print('Epoch', epoch)
            state = None
            for batch in tqdm.notebook.tqdm(train_iterator, leave=False):
                logits, state = self.network(batch.text, state)
                loss = F.cross_entropy(
                    logits.view(-1, logits.shape[-1]),
                    batch.target.view(-1).to(logits.device)
                )

                optimizer.zero_grad()
                loss.backward()
                optimizer.step()

            perplexity = self.dataset_perplexity(validation_dataset)
            print('Validation score:', perplexity)

            if perplexity < best_perplexity:
                best_perplexity = perplexity
                torch.save(self.network.state_dict(), './saves/best_LSTMModel.pt')

        self.network.load_state_dict(torch.load('./saves/best_LSTMModel.pt'))

    def next_word_probabilities(self, text_prefix):
        "Return a list of probabilities for each word in the vocabulary."

        prefix_token_tensor = torch.tensor(ids(text_prefix)).view(-1, 1)

        # YOUR CODE HERE

        with torch.no_grad():
            self.network.eval()

            logits, _ = self.network(prefix_token_tensor, None)
            probs = F.softmax(logits, dim=-1)[-1]

            self.network.train()

        return probs.squeeze(0).cpu().numpy()

    def dataset_perplexity(self, torchtext_dataset):
        "Return perplexity as a float."
        # Your code should be very similar to next_word_probabilities, but
        # run in a loop over batches. Use torch.no_grad() for extra speed.

        iterator = data.BPTTIterator(
            torchtext_dataset, batch_size=64, bptt_len=32, device=self.network.device)

        # YOUR CODE HERE

        log_probabilities = []
        with torch.no_grad():
            self.network.eval()

            state = None
            for batch in tqdm.notebook.tqdm(iterator, leave=False):
                logits, state = self.network(batch.text, state)
                probs = F.softmax(logits, dim=-1)

                selected_probs = probs.view(-1, probs.shape[-1])[
                    torch.arange(math.prod(batch.target.shape)),
                    batch.target.view(-1)
                ]
                log_probabilities.append(torch.log2(selected_probs).cpu().numpy())

            self.network.train()

        return 2 ** -np.mean(np.concatenate(log_probabilities))

lstm_model = LSTMModel(device)
lstm_model.train()

print('lstm validation perplexity:', lstm_model.dataset_perplexity(validation_dataset))
save_truncated_distribution(lstm_model, './saves/lstm_predictions.npy', short=False)

Epoch 0


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 500.3523832578693
Epoch 1


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 292.21037849436226
Epoch 2


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 231.046485959712
Epoch 3


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 201.85186824818555
Epoch 4


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 182.0928089954156
Epoch 5


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 170.37355394804797
Epoch 6


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 162.93251214521462
Epoch 7


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 156.31572539839425
Epoch 8


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 150.6057921773518
Epoch 9


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 146.57634451032615
Epoch 10


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 143.00659379625995
Epoch 11


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 142.2205021776439
Epoch 12


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 140.97854567534552
Epoch 13


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 137.44946240296332
Epoch 14


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 136.1639040175681
Epoch 15


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 136.17772115114929
Epoch 16


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 133.28160545006162
Epoch 17


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 131.99248669312786
Epoch 18


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 130.61844779918877
Epoch 19


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 129.69182979063822


  0%|          | 0/107 [00:00<?, ?it/s]

lstm validation perplexity: 129.69182979063822


  0%|          | 0/5000 [00:00<?, ?it/s]

saved ./saves/lstm_predictions.npy


<!-- Do not remove this comment, it is used by the autograder: RqYJKsoTS6 -->

Fill in your LSTM perplexity. 

LSTM validation perplexity: ***129.69182979063822***

# Experimentation: 1-Page Report

Now it's time for you to experiment.  Try to reach a validation perplexity below 120. You may either modify the LSTM class above, or copy it down to the code cell below and modify it there. Just **be sure to run code cell below to generate results with your improved LSTM**.  

It is okay if the bulk of your improvements are due to hyperparameter tuning (such as changing number or sizes of layers), but implement at least one more substantial change to the model.  Here are some ideas (several of which come from https://arxiv.org/pdf/1708.02182.pdf):
* activation regularization - add a l2 regularization penalty on the activation of the LSTM output (standard l2 regularization is on the weights)
* weight-drop regularization - apply dropout to the weight matrices instead of activations
* learning rate scheduling - decrease the learning rate during training
* embedding dropout - zero out the entire embedding for a random set of words in the embedding matrix
* ensembling - average the predictions of several models trained with different initialization random seeds
* temporal activation regularization - add l2 regularization on the difference between the LSTM output activations at adjacent timesteps

You may notice that most of these suggestions are regularization techniques.  This dataset is considered fairly small, so regularization is one of the best ways to improve performance.

For this section, you will submit a write-up describing the extensions and/or modifications that you tried.  Your write-up should be **1-page maximum** in length and should be submitted in PDF format.  You may use any editor you like, but we recommend using LaTeX and working in an environment like Overleaf.
For full credit, your write-up should include:
1.   A concise and precise description of the extension that you tried.
2.   A motivation for why you believed this approach might improve your model.
3.   A discussion of whether the extension was effective and/or an analysis of the results.  This will generally involve some combination of tables, learning curves, etc.
4.   A bottom-line summary of your results comparing validation perplexities of your improvement to the original LSTM.
The purpose of this exercise is to experiment, so feel free to try/ablate multiple of the suggestions above as well as any others you come up with!
When you submit the file, please name it `report.pdf`.



Run the cell below in order to train your improved LSTM and evaluate it.  

In [20]:
## Feel free to copy your original LSTM solution down here to modify for your report if you'd like.

class LSTMNetwork(Module):
    # a PyTorch Module that holds the neural network for your model

    def __init__(self, device):
        super().__init__()

        # YOUR CODE HERE

        self.lstm_network = nn.LSTM(128, 512, num_layers=3, dropout=0.5)
        self.hidden_network = nn.Sequential(
            nn.Dropout(0.5),
            nn.Linear(512, 128)
        )
        self.final_layer = nn.Linear(128, len(vocab))

        self.to_device(device)

    def forward(self, x, state):
        """Compute the output of the network.
        
        Note: In the Pytorch LSTM tutorial, the state variable is named "hidden":
        https://pytorch.org/tutorials/beginner/nlp/sequence_models_tutorial.html

        The torch.nn.LSTM documentation is quite helpful:
        https://pytorch.org/docs/stable/nn.html#lstm
    
        x - a tensor of int64 inputs with shape (seq_len, batch)
        state - a tuple of two tensors with shape (num_layers, batch, hidden_size)
                representing the hidden state and cell state of the of the LSTM.
        returns a tuple with two elements:
          - a tensor of log probabilities with shape (seq_len, batch, vocab_size)
          - a state tuple returned by applying the LSTM.
        """

        # Note that the nn.LSTM module expects inputs with the sequence 
        # dimension before the batch by default.
        # In this case the dimensions are already in the right order, 
        # but watch out for this since sometimes people put the batch first

        # YOUR CODE HERE

        x, state = self.move_inputs(x, state)

        x = F.embedding(x, self.final_layer.weight)
        x, state = self.lstm_network(x, state)
        return self.final_layer(self.hidden_network(x)), tuple([s.detach() for s in state])


class LSTMModel:
    "A class that wraps LSTMNetwork to handle training and evaluation."

    def __init__(self, device):
        self.network = LSTMNetwork(device)

    def train(self, n_epochs=20):
        train_iterator = data.BPTTIterator(
            train_dataset, batch_size=64, bptt_len=32, device=self.network.device)
        # Iterate over train_iterator with a for loop to get batches
        # each batch object has a .text and .target attribute with
        # token id tensors for the input and output respectively.

        # The initial state passed into the LSTM should be set to zero.

        # YOUR CODE HERE

        optimizer = torch.optim.Adam(self.network.parameters(), lr=1e-3, weight_decay=1e-5)

        best_perplexity = float('inf')
        for epoch in range(n_epochs):
            print('Epoch', epoch)
            state = None
            for batch in tqdm.notebook.tqdm(train_iterator, leave=False):
                logits, state = self.network(batch.text, state)
                loss = F.cross_entropy(
                    logits.view(-1, logits.shape[-1]),
                    batch.target.view(-1).to(logits.device)
                )

                optimizer.zero_grad()
                loss.backward()
                optimizer.step()

            perplexity = self.dataset_perplexity(validation_dataset)
            print('Validation score:', perplexity)

            if perplexity < best_perplexity:
                best_perplexity = perplexity
                torch.save(self.network.state_dict(), './saves/best_LSTMModel.pt')

        self.network.load_state_dict(torch.load('./saves/best_LSTMModel.pt'))

    def next_word_probabilities(self, text_prefix):
        "Return a list of probabilities for each word in the vocabulary."

        prefix_token_tensor = torch.tensor(ids(text_prefix)).view(-1, 1)

        # YOUR CODE HERE

        with torch.no_grad():
            self.network.eval()

            logits, _ = self.network(prefix_token_tensor, None)
            probs = F.softmax(logits, dim=-1)[-1]

            self.network.train()

        return probs.squeeze(0).cpu().numpy()

    def dataset_perplexity(self, torchtext_dataset):
        "Return perplexity as a float."
        # Your code should be very similar to next_word_probabilities, but
        # run in a loop over batches. Use torch.no_grad() for extra speed.

        iterator = data.BPTTIterator(
            torchtext_dataset, batch_size=64, bptt_len=32, device=self.network.device)

        # YOUR CODE HERE

        log_probabilities = []
        with torch.no_grad():
            self.network.eval()

            state = None
            for batch in tqdm.notebook.tqdm(iterator, leave=False):
                logits, state = self.network(batch.text, state)
                probs = F.softmax(logits, dim=-1)

                selected_probs = probs.view(-1, probs.shape[-1])[
                    torch.arange(math.prod(batch.target.shape)),
                    batch.target.view(-1)
                ]
                log_probabilities.append(torch.log2(selected_probs).cpu().numpy())

            self.network.train()

        return 2 ** -np.mean(np.concatenate(log_probabilities))

lstm_model = LSTMModel(device)
lstm_model.train()

print('lstm validation perplexity:', lstm_model.dataset_perplexity(validation_dataset))

Epoch 0


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 388.128721971458
Epoch 1


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 246.01421459753004
Epoch 2


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 206.94639789311802
Epoch 3


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 176.27944711380493
Epoch 4


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 167.6109693194254
Epoch 5


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 160.00713152561826
Epoch 6


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 147.3429682285497
Epoch 7


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 146.80975000553693
Epoch 8


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 136.43510182376102
Epoch 9


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 133.71388252808487
Epoch 10


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 133.95385488138479
Epoch 11


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 131.83596246217627
Epoch 12


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 129.57507381569113
Epoch 13


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 125.87144216810002
Epoch 14


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 124.47811153190679
Epoch 15


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 123.32985155719193
Epoch 16


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 122.0704783929067
Epoch 17


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 122.09202534182423
Epoch 18


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 120.35182560488327
Epoch 19


  0%|          | 0/1020 [00:00<?, ?it/s]

  0%|          | 0/107 [00:00<?, ?it/s]

Validation score: 120.91836738666359


  0%|          | 0/107 [00:00<?, ?it/s]

lstm validation perplexity: 120.35182560488327


### Submission

Upload a submission with the following files to Gradescope:
* hw1b.ipynb (rename to match this exactly)
* lstm_predictions.npy (this should also include all improvements from your exploration)
* neural_trigram_predictions.npy
* bigram_predictions.npy
* report.pdf

You can upload files individually or as part of a zip file, but if using a zip file be sure you are zipping the files directly and not a folder that contains them.

Be sure to check the output of the autograder after it runs.  It should confirm that no files are missing and that the output files have the correct format.  Note that the test set perplexities shown by the autograder are on a completely different scale from your validation set perplexities due to truncating the distribution and selecting different text.  Don't worry if the values seem much worse.